packages/block-library/src/template-part/edit/placeholder.js

```diff
@@ -0,0 +1,122 @@
+/**
+ * WordPress dependencies
+ */
+import { useEntityBlockEditor, EntityProvider } from '@wordpress/core-data';
+import { __ } from '@wordpress/i18n';
+import { BlockPreview } from '@wordpress/block-editor';
+import { useState, useCallback } from '@wordpress/element';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { cleanForSlug } from '@wordpress/editor';
+import { Placeholder, TextControl, Button } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import useTemplatePartPost from './use-template-part-post';
+
+function TemplatePartPreview() {
+	const [ blocks ] = useEntityBlockEditor( 'postType', 'wp_template_part' );
+	return (
+		<div className="wp-block-template-part__placeholder-preview">
+			<div className="wp-block-template-part__placeholder-preview-title">
+				{ __( 'Preview' ) }
+			</div>
+			<BlockPreview blocks={ blocks } />
+		</div>
+	);
+}
+
+export default function TemplatePartPlaceholder( { setAttributes } ) {
+	const [ slug, _setSlug ] = useState();
+	const [ theme, setTheme ] = useState();
+	const [ help, setHelp ] = useState();
+
+	// Try to find an existing template part.
+	const postId = useTemplatePartPost( null, slug, theme );
+
+	// If found, get its preview.
+	const preview = useSelect(
+		( select ) => {
+			if ( ! postId ) {
+				return;
+			}
+			const templatePart = select( 'core' ).getEntityRecord(
+				'postType',
+				'wp_template_part',
+				postId
+			);
+			if ( templatePart ) {
+				return (
+					<EntityProvider kind="postType" type="wp_template_part" id={ postId }>
+						<TemplatePartPreview />
+					</EntityProvider>
+				);
+			}
+		},
+		[ postId ]
+	);
+
+	const setSlug = useCallback( ( nextSlug ) => {
+		_setSlug( nextSlug );
+		setHelp( cleanForSlug( nextSlug ) );
+	}, [] );
```

Comment:
Normally, in this case, I would add _setSlug and setHelp as dependencies althougth I'm not expecting these functions to change. I wonder if it is totally safe to not include them in the dependencies.

Reply:
React cannot change these setters, and it's become an implicit convention not to list them in dependencies. I've also seen them always be listed. I think not listing them makes things easier to read.

Build tools will handle this at some point, so I don't have a strong opinion.

Reply:
From now on, I will follow the same approach and not list them 👍

Reply:
I think we've been avoiding function/variable names with `_` :)

Reply:
Really? Why? We've merged quite a bit of code with them. It's a standard convention for unused/intermediate variables. Forcing a different variable name usually ends up awkwardly.

Reply:
For me, it becomes an issue when there are two variables which share effectively the same name, but hold references to different values (overhead in distinguishing those values) or, worse, have differing implementations such that the two do not share the same meaning. Here, we call a value as "set slug", which can either mean to actually set the slug, _or_ to set the slug _and_ set the help text.

I might prefer instead:

- Name the callback in a way which encompasses all of the expected effects, e.g. `setSlugAndHelp`
- Name the callback in a way which represents its intended use, e.g. `onSlugChange`
   - I don't often encourage this since it falls into a trap of its own in that it is not descriptive of what we expect to happen when the slug is changed.
- Don't name the callback at all, e.g. anonymous function assignment in the `TextControl` attribute
- Or, if possible, remove the `help` state altogether, if it could be computed on-the-fly as a derivation of the slug (similar to [this old guideline](http://web.archive.org/web/20160430110308/http://facebook.github.io/react/tips/props-in-getInitialState-as-anti-pattern.html))

Reply:
>Name the callback in a way which encompasses all of the expected effects, e.g. setSlugAndHelp

In this particular case, this makes the most sense.

>Name the callback in a way which represents its intended use, e.g. onSlugChange
>- I don't often encourage this since it falls into a trap of its own in that it is not descriptive of what we expect to happen when the slug is changed.

Agreed

>Don't name the callback at all, e.g. anonymous function assignment in the TextControl attribute

It would make the code harder to follow here.

>Or, if possible, remove the help state altogether, if it could be computed on-the-fly as a derivation of the slug (similar to this old guideline)

We would still need some state because `help` can also be an error message. We could just set the status instead and compute `help` on the fly, but that might be harder to read.

What are your thoughts on things like this?

```js
const option = options.find( _option => _option === 'someOption' )
```

Reply:
> What are your thoughts on things like this?
> 
> ```js
> const option = options.find( _option => _option === 'someOption' )
> ```

I think it works a little better, but it could still be prone to some confusion in that: We call the value an "option", but in each occurrence, it represents a different thing:

- Either the result of an `Array#find` iteration, presumably having some meaning based on the context in which it's used (unclear in this example).
- A candidate of the `Array#find` iteration, one of the options which could ultimately become the resulting value.

We could choose to make this distinction clear in the naming, even if it ends up being more verbose. Because otherwise, we leave it to the reader to try to decipher this difference, which we should seek to avoid if there's a simple alternative.

```js
const selectedOption = options.find( option => option.isSelected );
```

```js
const currentRoute = routes.find( routeCandidate => routeCandidate.isMatch( currentURL ) );
```

(The "Candidate" suffix in this latter example isn't strictly necessary and arguably redundant, though maybe in some better examples a similar prefix/suffix help to clarify how it is a value under consideration)

In some cases, I think the thought experiment might lead us to call it something entirely different altogether, more representative of how it ends up being used.

```js
const page = posts.find( post => post.type === 'page );
```

I don't recall specifically where the conversation took place, but there's a similar issue with how we deal with variable shadowing in `useSelect`, where something like what you're suggesting might be appropriate enough if even to consider that the selectors are often the means-to-an-end of some resulting value.

```js
const isActive = useSelect( ( select ) => {
    const { isActive: _isActive } = select( 'foo' );
    return _isActive();
} );
```

I know @ellatrix has alternatively worked around this by defining the selector separately:

```js
function selector( select ) {
    const { isActive } = select( 'foo' );
    return isActive();
}

const isActive = useSelect( selector );
```

(I'm not sure if this is exactly right, though maybe it is.. depending on `const` and if variable hoisting is considered as being "shadowed")

Reply:
I think I would also go with:

```js
const selectedOption = options.find( option => option.isSelected );
```

or 

```js
const foundOption = options.find( option => option.isSelected );
```

>I don't recall specifically where the conversation took place, but there's a similar issue with how we deal with variable shadowing in useSelect, where something like what you're suggesting might be appropriate enough if even to consider that the selectors are often the means-to-an-end of some resulting value.

Yes, we could also prefix it with "select" or "get".

>(I'm not sure if this is exactly right, though maybe it is.. depending on const and if variable hoisting is considered as being "shadowed")

It's not shadowing, but it makes code verbose if you need closures.

I think we should have lint rules for these things. It's definitely inconsistent across the codebase.

Reply:
> I think we should have lint rules for these things. It's definitely inconsistent across the codebase.

For which, specifically?

As a baseline, we already have [`no-shadow`](https://eslint.org/docs/rules/no-shadow) enabled.

Regarding the underscore prefixing, I'm not sure we want to go so far as to forbid it, though I think in the course of this discussion it's become clearer that there are often better choices which can be made.

One potential interoperability concern would be with TypeScript checking, since it actually has special treatment of the underscore prefix when considering unused arguments (https://github.com/microsoft/TypeScript/issues/9458, https://github.com/microsoft/TypeScript/pull/9464, https://github.com/microsoft/TypeScript/issues/29202). (Aside: I don't love how this works, but it is what it is)

Or were you thinking about something specific to the selector callback for `useSelect` ?

Reply:
>For which, specifically?

- Prohibit bypassing `no-shadow` with an underscore prefix.
- Enforce "select" prefix when shadowing a variable with a selector.